builder/docker/step_build.go

```diff
@@ -22,28 +20,60 @@ func (s *stepBuild) Run(ctx context.Context, state multistep.StateBag) multistep
 
 	ui := state.Get("ui").(packer.Ui)
 	driver := state.Get("driver").(Driver)
-	ui.Say("Building base image...")
-
-	imageId, err := driver.Build(s.buildArgs.BuildArgs())
-	if err != nil {
+	config, ok := state.Get("config").(*Config)
+	if !ok {
+		err := fmt.Errorf("error encountered obtaining docker config")
 		state.Put("error", err)
+		ui.Error(err.Error())
 		return multistep.ActionHalt
 	}
 
-	ui.Sayf("Finished building base image %q", imageId)
+	ui.Say("Building base image...")
 
-	cfg, ok := state.GetOk("config")
-	if !ok {
-		state.Put("error", errors.New("missing config in state; this is a docker plugin bug, please report upstream"))
-		return multistep.ActionHalt
+	if config.EcrLogin {
```

Comment:
I'm assuming this code is quite similar, if not equivalent, to the code from `step_pull.go`, is that right?
If so, I would suggest abstracting this logic (and cleanup function) to a common method, so that we don't duplicate this code.

I see it is already duplicated also for the post-processors when applicable, so I would understand if you prefer to continue with this for now, we can always reify this function in a later PR as well.

Reply:
> I'm assuming this code is quite similar, if not equivalent, to the code from `step_pull.go`, is that right?

Yes, it is effectively copied and pasted (with minor conforming cleanup to grab the config from `state` and refactor some error messages.

> If so, I would suggest abstracting this logic (and cleanup function) to a common method, so that we don't duplicate this code.
>
> I see it is already duplicated also for the post-processors when applicable, so I would understand if you prefer to continue with this for now, we can always reify this function in a later PR as well.

I am all in favor of abstracting this logic to keep the code DRY, but was trying to keep my code changes to a minimum so as to make review and approval easier / faster, as the lack of support for login before a bootstrapped build is currently preventing us from using Packer with our configuration. I also didn't want to inadvertently break something, since I am new to this code!

If this otherwise looks OK and you are comfortable approving, I would suggest we keep the duplicated code for the time being.  In the meantime I can begin working on a new PR to abstract the the logic, but am guessing that'll take me a bit longer.

Reply:
Oh yeah no pressure on this PR specifically, the current code looks good to me, just wanted to point-out that we could mutualise that logic, but as I mentioned and as you point out as well, this can be done later.

I can probably take care of this once your PRs have been integrated, that way we can proceed with releasing a new version of the plugin next week as well so those changes are usable then.